qiime2/sdk/parallel_context.py

```diff
@@ -0,0 +1,247 @@
+            if isinstance(proxy, Proxy):
+                futures.append(proxy._future_)
+                mapped.append(proxy.__class__(
+                    len(futures) - 1, proxy._selector_))
+            else:
+                mapped.append(proxy)
+    elif isinstance(arg, dict):
+        mapped = {}
+
+        for key, value in arg.items():
+            if isinstance(value, Proxy):
+                futures.append(value._future_)
+                mapped[key] = value.__class__(
+                    len(futures) - 1, value._selector_)
+            else:
+                mapped[key] = value
+    # We just have a real artifact and don't need to map
+    else:
+        mapped = arg
+
+    return mapped
+
+
+def _unmap_arg(arg, inputs):
+    """ Unmap a proxy artifact given to a parsl action
+    """
+    # We were hacky and set _future_ to be the index of this artifact in the
+    # inputs list
+    if isinstance(arg, Proxy):
+        resolved_result = inputs[arg._future_]
+        unmapped = arg._get_element_(resolved_result)
+    # If we got a collection of proxies as the input we were even hackier and
+    # added each proxy to the inputs list individually while having a list of
+    # their indices in the args.
+    elif isinstance(arg, list):
+        unmapped = []
+
+        for proxy in arg:
+            if isinstance(proxy, Proxy):
+                resolved_result = inputs[proxy._future_]
+                unmapped.append(proxy._get_element_(resolved_result))
+            else:
+                unmapped.append(proxy)
+    elif isinstance(arg, dict):
+        unmapped = {}
+
+        for key, proxy in arg.items():
+            if isinstance(proxy, Proxy):
+                resolved_result = inputs[proxy._future_]
+                unmapped[key] = proxy._get_element_(resolved_result)
+            else:
+                unmapped[key] = proxy
+    # We didn't have a proxy at all
+    else:
+        unmapped = arg
+
+    return unmapped
+
+
+def _contains_proxies(*args, **kwargs):
+    """Returns True if any of the args or kwargs are proxies
+    """
+    return any(isinstance(arg, Proxy) for arg in args) \
+        or any(isinstance(value, Proxy) for
+               value in kwargs.values())
+
+
+class ParallelContext(Context):
+    def __init__(self, action_obj, parent=None):
+        super(ParallelContext, self).__init__(action_obj, parent=parent)
+
+        if parent is not None:
+            self.action_executor_mapping = parent.action_executor_mapping
+            self.executor_name_type_mapping = parent.executor_name_type_mapping
+        else:
+            self.action_executor_mapping = \
+                PARALLEL_CONFIG.action_executor_mapping
+            self.executor_name_type_mapping = \
+                None if PARALLEL_CONFIG.parallel_config is None \
+                else {v.label: v.__class__.__name__
+                      for v in PARALLEL_CONFIG.parallel_config.executors}
+
+    def callable_action(self, *args, **kwargs):
```

Comment:
This should probably be private, and since it would be called by the parent class, we might give it a special name like `_callable_action_()`.

Reply:
Oh. I don't think this needs to exist on this class right? The base-class appears to have the same implementation.

Reply:
The only difference is the implementation on ParallelContext short circuits if it receives proxies as args. The version on base Context doesn't check for Proxies so it can remain unaware of all parallel stuff.

Reply:
got it, ok then that's a good reason to override.


We should probably still change the name to `_callable_action_`